File: MATLAB/ExtendedYaleBData/CreateSubsets.m
%%
% Script that creates subsets from Extended Yale B database according to
% azimuth and elevation angles of single light source
% 5 increasingly complex (and more challenging) subsets. This segmentation
% of the database has been used in other publications as well, e.g. 
% Wright et al., 2009
% Image processing toolbox needed. Alternatively, the python implementation
% can be used.
% Part of RobOMP package. "RobOMP: Robust variants of Orthogonal Matching 
% Pursuit for sparse representations" DOI: 10.7717/peerj-cs.192 (open access)
% Author: Carlos Loza
% https://github.carlosloza/RobOMP
% Note: Assuming directories as in remote repo, downloaded (from 
% http://vision.ucsd.edu/~iskwak/ExtYaleDatabase/ExtYaleB.html) and unziped
% directory (Cropped Images version) should be placed in '../../Data'

close all
clearvars
clc

subj_v = [1:13 15:39];              % Subjects of database
nSub = length(subj_v);

fprintf('Creating subsets... \n')
for i = 1:nSub
    if subj_v(i) < 10
        DirFiles = ['../../Data/CroppedYale/yaleB0' num2str(subj_v(i))];
    else
        DirFiles = ['../../Data/CroppedYale/yaleB' num2str(subj_v(i))];
    end
    % Create directories for each subset
    for j = 1:5
        mkdir([DirFiles '/Subset' num2str(j)])
    end
    FileList = dir(fullfile(DirFiles, '*.pgm'));
    nImg = size(FileList,1);
    for j = 1:nImg
        auxImg = imread([DirFiles '/' FileList(j).name]);
        a1 = str2double(FileList(j).name(end-10:end-8));
        a2 = str2double(FileList(j).name(end-5:end-4));
        angl = round(sqrt(a1^2 + a2^2));
        if angl <= 12
            subs = '/Subset1/';
        elseif angl >= 13 && angl <= 25
            subs = '/Subset2/';
        elseif angl >= 26 && angl <= 54
            subs = '/Subset3/';
        elseif angl >= 55 && angl < 83
            subs = '/Subset4/';
        elseif angl >= 84
            subs = '/Subset5/';
        end
        imwrite(auxImg, [DirFiles subs FileList(j).name]);
    end
    sd = 1;
end
fprintf('Done! \n')

File: ExtendedYaleBData/CreateSubsets.py
"""
Script that creates subsets from Extended Yale B database according to
azimuth and elevation angles of single light source
5 increasingly complex (and more challenging) subsets. This segmentation
of the database has been used in other publications as well, e.g.
Wright et al., 2009
Part of RobOMP package. "RobOMP: Robust variants of Orthogonal Matching
Pursuit for sparse representations" DOI: 10.7717/peerj-cs.192 (open access)
Author: Carlos Loza
https://github.carlosloza/RobOMP
Note: Assuming directories as in remote repo, downloaded (from
http://vision.ucsd.edu/~iskwak/ExtYaleDatabase/ExtYaleB.html) and unziped
directory (Cropped Images version) should be placed in '../../Data'
"""

import os
import glob

# Subjects of database
subj_v = list(range(1, 14)) + list(range(15, 40))
nSub = len(subj_v)

print('Creating subsets...')
for i in range(0, nSub):
    if subj_v[i] < 10:
        DirFiles = '../Data/CroppedYale/yaleB0' + str(subj_v[i])
    else:
        DirFiles = '../Data/CroppedYale/yaleB' + str(subj_v[i])
    # Create directories for each subset
    for j in range(1, 6):
        path = DirFiles + '/Subset' + str(j)
        os.mkdir(path)
    FileList = [f for f in glob.glob(DirFiles + '/*.pgm')]
    nImg = len(FileList)
    for j in range(0, nImg):
        if FileList[j][-11:-4] == 'Ambient':
            print('Ignore file')
        else:
            a1 = int(FileList[j][-11:-8])
            a2 = int(FileList[j][-6:-4])
            angl = round((a1**2 + a2**2)**0.5)
            if angl <= 12:
                subs = '/Subset1/'
            elif angl >= 13 and angl <= 25:
                subs = '/Subset2/'
            elif angl >= 26 and angl <= 54:
                subs = '/Subset3/'
            elif angl >= 55 and angl <= 83:
                subs = '/Subset4/'
            elif angl >= 84:
                subs = '/Subset5/'

            os.rename(FileList[j], DirFiles + subs + FileList[j][len(DirFiles)+1:])
print('Done!')